RFCs/FS-1063-support-letbang-andbang-for-applicative-functors.md

```diff
@@ -0,0 +1,750 @@
+# F# RFC FS-1063 - Support let! .. and... for applicative functors
+
+The design suggestion [Support let! .. and... for applicative functors](https://github.com/fsharp/fslang-suggestions/issues/579) has been marked "approved in principle".
+This RFC covers the detailed proposal for this suggestion.
+
+* [x] [Approved in principle](https://github.com/fsharp/fslang-suggestions/issues/579#event-1345361104) & [prioritised](https://github.com/fsharp/fslang-suggestions/issues/579#event-1501977428)
+* [x] [Suggestion](https://github.com/fsharp/fslang-suggestions/issues/579)
+* [x] [Discussion](https://github.com/fsharp/fslang-design/issues/335)
+* [ ] Implementation: [In progress](https://github.com/Microsoft/visualfsharp/pull/5696)
+
+# Summary
+[summary]: #summary
+
+Extend computation expressions to support applicative functors via a new `let! ... and! ... return ...` syntax.
+
+# Motivation
+[motivation]: #motivation
+
+Applicative functors (or just "applicatives", for short) have been growing in popularity as a way to build applications and model certain domains over the last decade or so, since McBride and Paterson published [Applicative Programming with Effects](http://www.staff.city.ac.uk/~ross/papers/Applicative.html). Applicatives are now reaching a level of popularity within the community that supporting them with a convenient and readable syntax, as we do for monads, makes sense.
```

Comment:
I would argue it's not specifically applicatives but something along the lines of "at the same time," e.g. parallel, concurrent, etc. that we should push as the focal point. Applicatives are the category theory term, but computation expressions are not merely about category theory. I think more people will understand the desire for, e.g. in `async` computations, to await multiple, unrelated computations at the same time rather than _applicatives_ in particular. That said, references to applicatives are certainly useful as a means of showing why the implementation may follow a certain path.

Reply:
Yes, I did have this sort of thing in mind at some point and obviously failed to put it into the RFC at any point. I'll revisit this, because I think it's an important aspect.

Reply:
fwiw, I want to be careful to avoid baking in the `async` as the canonical example as the intuition for this - the feature is much more general than that.

Reply:
I've added an example of parallelism because, despite my fears, it is a very useful example of the power of applicatives. I've tried to emphasise the independence of arguments as part of that.